Tests/Functional/Segment/Query/Functional/QueryFunctionalTest.php

```diff
@@ -0,0 +1,217 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MauticPlugin\CustomObjectsBundle\Tests\Functional\Segment\Query\Functional;
+
+use Doctrine\ORM\Exception\ORMException;
+use Doctrine\ORM\OptimisticLockException;
+use Mautic\CoreBundle\Test\MauticMysqlTestCase;
+use Mautic\LeadBundle\Command\UpdateLeadListsCommand;
+use Mautic\LeadBundle\Entity\Lead;
+use Mautic\LeadBundle\Entity\LeadList;
+use Mautic\LeadBundle\Entity\ListLead;
+use Mautic\LeadBundle\Provider\FilterOperatorProviderInterface;
+use MauticPlugin\CustomObjectsBundle\CustomFieldType\TextType;
+use MauticPlugin\CustomObjectsBundle\Entity\CustomField;
+use MauticPlugin\CustomObjectsBundle\Entity\CustomFieldValueText;
+use MauticPlugin\CustomObjectsBundle\Entity\CustomItem;
+use MauticPlugin\CustomObjectsBundle\Entity\CustomItemXrefContact;
+use MauticPlugin\CustomObjectsBundle\Entity\CustomObject;
+use MauticPlugin\CustomObjectsBundle\Model\CustomItemModel;
+use MauticPlugin\CustomObjectsBundle\Provider\ConfigProvider;
+use Symfony\Component\Translation\TranslatorInterface;
+
+class QueryFunctionalTest extends MauticMysqlTestCase
+{
+    private CustomField $customField;
+
+    private Lead $leadABC;
+
+    private Lead $leadXYZ;
+
+    private Lead $leadEmpty;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->configParams[ConfigProvider::CONFIG_PARAM_ITEM_VALUE_TO_CONTACT_RELATION_LIMIT] = 0;
+
+        $customItemModel       = self::$container->get('mautic.custom.model.item');
+        $this->assertInstanceOf(CustomItemModel::class, $customItemModel);
+
+        [$customObject, $this->customField] = $this->createBookObjectWithPublisherCustomField();
+
+        $values =  ['bookWithPublisherABC' => 'ABC', 'bookWithPublisherXYZ' => 'XYZ', 'bookWithPublisherEmpty' => ''];
+
+        foreach ($values as $name => $publisher) {
+            ${$name} = $this->createCustomItem($customObject, $publisher, 'publisher'.$publisher);
+        }
+
+        $this->leadABC = new Lead();
+        $this->leadABC->setFirstname('LeadABC')->setEmail('leadABC@acquia.com')->setIsPublished(true);
+        $this->em->persist($this->leadABC);
+
+        $this->leadXYZ = new Lead();
+        $this->leadXYZ->setFirstname('LeadXYZ')->setEmail('leadXYZ@acquia.com')->setIsPublished(true);
+        $this->em->persist($this->leadXYZ);
+
+        $this->leadEmpty = new Lead();
+        $this->leadEmpty->setFirstname('LeadEmpty')->setEmail('leadEmpty@acquia.com')->setIsPublished(true);
+        $this->em->persist($this->leadEmpty);
+
+        $this->em->flush();
+
+        /** @phpstan-ignore-next-line */
+        $xrefABC = $customItemModel->linkEntity($bookWithPublisherABC, 'contact', (int) $this->leadABC->getId());
```

Comment:
Yes, phpstan was rejecting this line since I am defining those variables in the foreach loop. I used that because defining everything again and again seemed tedious. It does make it less readable though, I agree. 

I'll define them without the foreach